### DISCUSSION TITLE
Shared library build for AIX

### DISCUSSION THREAD

**Discussion by sr-kuantum** [mesonbuild/meson#10528] [Q&A]
Hello team,

I'm trying to build one of the package using meson. The package generates both shared and static library.
But as part of the AIX linking, we want the static library to archive the ".so" file instead of the ".o" files.

I was trying to look in the code where this shared library is being generated. So that once the .so is generated, I would like to archive it to a .a.

I would really appreciate it if you could show me some direction.

**Comment by dcbaker** [comment]
That's rather unusual. On Linux ar doesn't even understand .so formats (they're executable), usually you would do it the other way, create an archive, and then create a shared library from that. you would do that with `shared_library('foo', link_whole : <static_library>)`.

Due to a bug you *can* write `static_library('foo', link_with : shared_library(...))`, but it will result in an empty .a file. So you'd need a feature request for that.

What are you trying to achieve by archiving the shared library that wouldn't be achieved by creating a shared_library from the static archive?

**Reply by ayappanec** [reply]
In AIX, generally the shared libraries are shipped inside archives. This has some benefits for a server operating system like AIX since the older/legacy applications in AIX using the older shared objects can still use it and other new applications can be made to use the newer shared objects.
For example, let's take openssl library (libssl) in AIX. 
$ ar -tv /usr/lib/libssl.a
rwxr-xr-x     0/0     728648 Oct 21 11:44 2021 libssl.so
rwxr-xr-x     0/0     510766 Oct 21 11:44 2021 libssl.so.0.9.8
rwxr-xr-x     0/0     728648 Oct 21 11:44 2021 libssl.so.1.0.0
rwxr-xr-x     0/0     728648 Oct 21 11:44 2021 libssl.so.1.0.2
rwxr-xr-x     0/0     1061779 Oct 21 11:37 2021 libssl.so.1.1

It has several shared objects inside the archive. Older applications linking to "libssl.a(libssl.so.1.0.2)" continue to work alongside other applications which links to the newer libssl.a(libssl.so.1.1).
Having an archive kind of mechanism for shared objects also helps in avoiding the problem of recompiling the applications which links to the shared library in the case where the shared library is updated.

**Reply by sr-kuantum** [reply]
@dcbaker, thanks for responding.
AIX static library linking is different from Linux, the .so is shipped as part of the static library. We are trying to build a similar way with glib2 library to ship the .so in the static library.


Below is an example.

>> ar -X64 -tv /usr/lib/libcrypto.a
rwxr-xr-x     0/0     3488235 Oct 21 11:44 2021 libcrypto.so
rwxr-xr-x     0/0     2527801 Oct 21 11:44 2021 libcrypto.so.0.9.8
rwxr-xr-x     0/0     3488235 Oct 21 11:44 2021 libcrypto.so.1.0.0
rwxr-xr-x     0/0     3488235 Oct 21 11:44 2021 libcrypto.so.1.0.2
rwxr-xr-x     0/0     5107323 Oct 26 02:13 2021 libcrypto.so.1.1
rwxr-xr-x     0/0     3488235 Oct 21 11:44 2021 libcrypto64.so
rwxr-xr-x     0/0     2527801 Oct 21 11:44 2021 libcrypto64.so.0.9.8
rwxr-xr-x     0/0     3488235 Oct 21 11:44 2021 libcrypto64.so.1.0.0

**Reply by dcbaker** [reply]
This is definately something we don't currently support, and will need some kind of Meson DSL support for, I've opend #10532 to dicuss that further

**Comment by sr-kuantum** [comment]
@dcbaker, could you please provide some pointers on how the static libs are build by the meson code? For the below mentioned  examples
https://mesonbuild.com/Build-targets.html

lib = static_library()

Is there any-other way we can mention in the meson.build file to build the static libs?

-Ddefault_library=static ??

**Comment by sr-kuantum** [comment]
@dcbaker  Can this code change help here?

The rule in the build.ninja file updated to pick the static library instead of the object files.

```
build liblibcall.a: STATIC_LINKER liblibcall.so
  LINK_ARGS = -csr -Xany 
```

The build generated the static lib with the archived shared lib

 

```

> 
> $ ninja -v
> [1/5] gcc -Iliblibcall.so.p -I. -I.. -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -fPIC -MD -MQ liblibcall.so.p/qall.c.o -MF liblibcall.so.p/qall.c.o.d -o liblibcall.so.p/qall.c.o -c ../qall.c
> [2/5] gcc -Iliblibcall.so.p -I. -I.. -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -fPIC -MD -MQ liblibcall.so.p/call.c.o -MF liblibcall.so.p/call.c.o.d -o liblibcall.so.p/call.c.o -c ../call.c
> [3/5] gcc -Iliblibcall.so.p -I. -I.. -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -fPIC -MD -MQ liblibcall.so.p/kall.c.o -MF liblibcall.so.p/kall.c.o.d -o liblibcall.so.p/kall.c.o -c ../kall.c
> [4/5] gcc  -o liblibcall.so liblibcall.so.p/call.c.o liblibcall.so.p/kall.c.o liblibcall.so.p/qall.c.o -Wl,-bernotok -Wl,-bnoipath -Wl,-bbigtoc -shared -fPIC -Wl,-blibpath:/opt/freeware/lib/gcc/powerpc-ibm-aix7.2.0.0/8:/opt/freeware/lib:/lib:/usr/lib
> [5/5] ar -csr -Xany liblibcall.a liblibcall.so
> 
>
> 
> $ ar -t liblibcall.a
> liblibcall.so

```


Code changes
```

--- ninjabackend.py     2022-09-07 11:37:11 +0000
+++ ninjabackend.py.bk  2022-09-07 11:56:22 +0000
@@ -3003,6 +3003,9 @@
         implicit_outs = []
         if isinstance(target, build.StaticLibrary):
             linker_base = 'STATIC'
+            obj_list.clear()
+            obj_list.append(outname[:-1]+'so')
         else:
             linker_base = linker.get_language() # Fixme.
         if isinstance(target, build.SharedLibrary):
```
